src/odbc/unittests/connection_string_parse.c

```diff
@@ -0,0 +1,198 @@
+#include "common.h"
+#include <assert.h>
+#include "freetds/odbc.h"
+
+
+#ifdef _WIN32
+HINSTANCE hinstFreeTDS;
+#endif
+
+static void
+assert_equal_str(DSTR a, const char *b)
+{
+  assert(b && strncmp(tds_dstr_cstr(&a), b, tds_dstr_len(&a))==0);
+}
+
+static int
+simple_string(void)
+{
+  TDSLOGIN *login;
+  TDS_ERRS errs = {0};
+  TDSLOCALE *locale;
+  TDS_PARSED_PARAM parsed_params[ODBC_PARAM_SIZE];
+  const char *connect_string = "DRIVER=libtdsodbc.so;SERVER=127.0.0.1;PORT=1337;UID=test_username;PWD=test_password;DATABASE=test_db;ClientCharset=UTF-8;";
+
+  const char *connect_string_end = connect_string + strlen(connect_string);
+  login = tds_alloc_login(0);
+  tds_set_language(login, "us_english");
+  locale = tds_alloc_locale();
+  login = tds_init_login(login, locale);
+
+  odbc_errs_reset(&errs);
+
+  if (!odbc_parse_connect_string(&errs, connect_string, connect_string_end, login, parsed_params))
+     return 1;
+
+  assert_equal_str(parsed_params[ODBC_PARAM_UID], "test_username");
+  assert_equal_str(login->server_name, "127.0.0.1");
+  assert_equal_str(login->password, "test_password");
+  assert_equal_str(parsed_params[ODBC_PARAM_PWD], "test_password");
+  assert(login->port == 1337);
+
+  tds_free_login(login);
+  tds_free_locale(locale);
+
+  return 0;
+}
+
+static int
+simple_escaped_string(void)
+{
+  TDSLOGIN *login;
+  TDS_ERRS errs = {0};
+  TDSLOCALE *locale;
+  TDS_PARSED_PARAM parsed_params[ODBC_PARAM_SIZE];
+  const char *connect_string = "DRIVER={libtdsodbc.so};SERVER={127.0.0.1};PORT={1337};UID={test_username};PWD={test_password};DATABASE={test_db};ClientCharset={UTF-8};";
+
+  const char *connect_string_end = connect_string + strlen(connect_string);
+  login = tds_alloc_login(0);
+  tds_set_language(login, "us_english");
+  locale = tds_alloc_locale();
+  login = tds_init_login(login, locale);
+
+  odbc_errs_reset(&errs);
+
+  if (!odbc_parse_connect_string(&errs, connect_string, connect_string_end, login, parsed_params))
+     return 1;
+
+  assert_equal_str(parsed_params[ODBC_PARAM_UID], "test_username");
+  assert_equal_str(login->server_name, "127.0.0.1");
+  assert_equal_str(login->password, "test_password");
+  assert_equal_str(parsed_params[ODBC_PARAM_PWD], "test_password");
+  assert(login->port == 1337);
+
+  tds_free_login(login);
+  tds_free_locale(locale);
+
+  return 0;
+}
+
+static int
+test_special_symbols(void)
+{
+  TDSLOGIN *login;
+  TDS_ERRS errs = {0};
+  TDSLOCALE *locale;
+  TDS_PARSED_PARAM parsed_params[ODBC_PARAM_SIZE];
+  const char *connect_string = "DRIVER={libtdsodbc.so};SERVER={127.0.0.1};PORT={1337};UID={test_username};PWD={[]{}(),;?*=!@};DATABASE={test_db};ClientCharset={UTF-8};";
```

Comment:
I tested with native Microsoft drivers. The `}` need to be quoted so if you want `PWD` to be `[]{}(),;?*=!@` the syntax is `PWD={[]{}}(),;?*=!@}` otherwise with `PWD={[]{}(),;?*=!@}` you will get `Invalid connection string attribute` error